src/profile/ProfileForm.tsx

```diff
@@ -37,585 +23,315 @@ interface ChangedUser {
-      case "password": {
-        let newErrors: string[] = [];
-
-        if (value.length === 0) {
-          dispatch({
-            type: "set data",
-            data: { password: undefined },
-            errors: { status_code: 0, password: [] },
-          });
-          break;
-        }
-
-        if (value.length < 8) {
-          newErrors.push("USER.PASSWORD.TOO_SHORT.8");
-        }
+  formId?: string;
+  context: FormContextValues<ProfileFormVals>;
+  onSubmit?: (
+    vals: ProfileFormVals,
+    e?: React.BaseSyntheticEvent
+  ) => Promise<void>;
+  hideSubmit?: boolean;
+}
 
-        if (/^\d*$/.test(value)) {
-          newErrors.push("USER.PASSWORD.ENTIRELY_NUMERIC");
-        }
+const ProfileForm: React.FC<ProfileFormProps> = (props: ProfileFormProps) => {
+  const auth = useAuth();
 
-        dispatch({
-          type: "set data",
-          data: { password: value },
-          errors: { status_code: 0, password: newErrors },
-        });
-        break;
+  // react-hook-form caches default values when the first time you call useForm.
+  // So, we have to manually replicate the defaultValues behaviour by resetting the form values ourselves
+  // whenever user changes
+  const { reset, register } = props.context;
+
+  useEffect(() => {
+    reset({
+      username: props.user?.username,
+      first_name: props.user?.first_name,
+      last_name: props.user?.last_name,
+      email: props.user?.email,
+      is_editor: props.user?.is_editor || false,
+      writer_name: props.user?.writer_name,
+    });
+  }, [props.user, reset]);
+
+  // manually register the is_editor field since we handle it with a select
+  useEffect(() => {
+    register({
+      name: "is_editor",
+    });
+  }, [register]);
+
+  const [showPassword, setShowPassword] = useState<boolean>(false);
+
+  const submit = async (vals: ProfileFormVals) => {
+    const cleaned = cleanFormData(vals);
+
+    const editingMe = props.user?.username === auth.user?.username;
+    try {
+      if (props.user === undefined) {
+        await auth.post<ProfileFormVals>("users/", cleaned);
+      } else {
+        await auth.patch<ProfileFormVals>(
+          editingMe ? "me/" : `users/${vals?.username}/`,
+          cleaned,
+          editingMe
+        );
       }
+    } catch (err) {
+      const apiErrors = err as UserAPIError;
+      props.context.setError(
+        apiErrors.cur_password?.map((error) => ({
+          name: "cur_password",
+          type: "server_error",
+          message: error,
+        }))
+      );
     }
-  };
 
-  const onSubmit = (evt?: React.FormEvent<HTMLFormElement>) => {
-    evt?.preventDefault();
-
-    if (
-      (Object.keys(state.changedUser) as Array<keyof ChangedUser>).every(
-        (k) =>
-          state.changedUser[k] === undefined ||
-          (k === "is_editor" && state.changedUser[k] === user?.is_editor)
-      )
-    ) {
-      return Promise.resolve();
+    if (props.onSubmit) {
+      await props.onSubmit(vals);
```

Comment:
onSubmit has no way of knowing if form submission failed or not. This seems like an oversight.

Reply:
My mastermind idea was that if an exception was thrown, it would catch and not call the onSubmit code, which would have worked if the call was actually inside the try block.

https://github.com/UWmathNEWS/slugline-web/pull/13/commits/0ae171358513d97053e8b78183be48cda056e5ed